### ISSUE TITLE
[Feature Request] Docker Auth from Vault

### ISSUE THREAD

**Issue by the-maldridge** [hashicorp/nomad#9740]
Right now the options for auth in the docker driver are pretty much hard-coded credentials or credential helpers.  For repos that just take basic auth this is a bit overkill.  It would be nice to be able to specify a vault key to retrieve credentials from.  It could be a hard-coded schema within the key, so you just specify something like:

```
auth {
  vault = "secret/docker/ghcr"
}
```

In theory this shouldn't affect the scheduler since vault can be assumed to be a cluster service if its configured at all.

**Comment by tgross** [comment]
Hi @the-maldridge! I'm 👍 on this feature for sure. We have a similar issue open already https://github.com/hashicorp/nomad/issues/4576 but I think this is a distinct enough request that we should keep it open and cross-link it for discussion. Thanks for opening this!

**Comment by krishicks** [comment]
:wave: I wrote up an RFC for implementing something like this, but after getting feedback on it I'm going to suggest an alternative for the moment, as existing functionality can be used to achieve this, as demonstrated by @angrycub:

```hcl
job "auth" {
  type        = "service"
  datacenters = ["dc1"]

  group "docker" {
    task "redis" {
      driver = "docker"

      config {
        image = "registry.service.consul:5000/redis:latest"
        auth {
          username = "${DOCKER_USER}"
          password = "${DOCKER_PASS}"
        }
      }

      template {
        destination = "secrets/secret.env"
        env         = true
        change_mode = "restart"
        data        = <<EOH
DOCKER_USER={{ key "kv/docker/config/user" }}
DOCKER_PASS={{ key "kv/docker/config/pass" }}
EOH
      }

      resources {
        cpu    = 200
        memory = 100
      }
    }
  }
}
```

This uses the `template` stanza to render the Docker credentials from Vault as environment variables, which are then interpolated into the Docker driver config.

This solution assumes that you're OK with having the Docker credentials made available to the task (as they would exist as environment variables the task can read).

Let me know if this solves the issue, or if you have different requirements that make it not suitable.

**Comment by the-maldridge** [comment]
While this does allow you to pass the docker credentials via the environment, it exposes the credentials to the task as you mentioned.  Since this fails to meet the criteria of least privilege, this fails to serve as an equivalent solution.

**Comment by tgross** [comment]
I've re-titled this issue to cover Variables as well, and I'm going to close some of the duplicate requests for this feature and point them here, as well as make sure this gets into Jira for roadmapping discussions.

**Comment by MaksSieve** [comment]
any updates on this issue?

**Comment by mismithhisler** [comment]
Hi @MaksSieve! This is now possible using the [Secret block](https://developer.hashicorp.com/nomad/docs/job-specification/secret) which was added as a feature in 1.11.0.

An example of using this would be:
```
job "docs" {
  group "example" {
    task "server" {

      secret "docker_auth" {
        provider = "vault"
        path     = "path/to/secret"
        config {
          engine = "kv_v2"
        }
      }

      driver = "docker"
      
      config {
        image = "secret/service"
    
        auth { 
          username = "${secret.docker_auth.username}"
          password = "${secret.docker_auth.password}"
        }
      }

    }
  }
}
```

The Vault secret would need the keys `username` and `password`. I'll update our Docker driver auth registry section to make note of using the Secret block.

I'm going to close this issue out as completed. If you have any troubles using the Secret block, feel free to open up another issue.

**Comment by the-maldridge** [comment]
@mismithhisler while you're updating docs, could you make a pass over the artifact page in the job specification and make a note about whether or not secrets are supported in there?  There's a good note in that page already about unveiling environment variables which may originate from other sources such as secrets, but it would be great to have concrete documentation about secrets being exposed or not.

**Comment by mismithhisler** [comment]
@the-maldridge, absolutely! The artifact parameters/blocks `source` `destination` `headers {}` and `options {}` can all be interpreted/replaced with a values from the Secret block. This is all done by the Nomad client process.

The go-getter subprocess will look for certain environment variables (ex. aws secret/key if the source is an s3 source). We keep this subprocess sandboxed and only give it access to the envvars you **explicitly** allow (via `artifact.set_environment_variables`, or via the artifact block params).

I see how these docs are confusing and will get this tidied up. Thanks for the feedback!
